draft-ietf-quic-tls.md

```diff
@@ -580,6 +661,185 @@ Prior to TLS providing keys, no record protection is performed and the
+messages sent by an endpoint MUST NOT be sent in protected packets,
+specifically:
+
+* A client MUST NOT restransmit a TLS ClientHello with 0-RTT keys.  The server
+  needs this message in order to determine the 0-RTT keys.
+
+* A server MUST NOT retransmit any of its TLS handshake messages with 1-RTT
+  keys.  The client needs these messages in order to determine the 1-RTT keys.
+
+A HelloRetryRequest might be used to reject an initial ClientHello.  A
+HelloRetryRequest handshake message and any second ClientHello that is sent in
+response MUST also be sent without packet protection.  This is natural, because
+no new keying material will be available when these messages need to be sent.
+
+Note:
+
+: An alternative way of identifying handshake data that needs to be sent without
+  protection is to collect all handshake data from before TLS provides the first
+  keys (see {{key-ready-events}}).
+
+Retransmissions of these handshake messages MUST be send in unprotected packets
+(with a KEY_PHASE of 0).  Any ACK frames for these messages MUST also be send in
+unprotected packets.
+
+
+### Handshake Retransmission and 0-RTT
+
+The first protected packets sent by a client that has not attempted to send
+0-RTT data or packets sent by a server will be marked with a KEY_PHASE of 1.
+This distinguishes those protected packets from unprotected packets.  Though loss
+or reordering might cause unprotected packets to arrive once 1-RTT keys are in
+use, unprotected packets are easily distinguished from 1-RTT packets.  Based on
+the KEY_PHASE bit, these packets might initially appear to be a key update (see
+{{key-update}}), but they will not be decrypted successfully and so will be
+discarded.
+
+However, if a client has sent 0-RTT data, unprotected packets will not be
+clearly distinguished from 1-RTT protected packets.  Both will have a KEY_PHASE
+of 0.  A server that does not yet have 1-RTT keys will correctly interpret these
+as being unprotected, but a server that has 1-RTT keys could reject unprotected
+packets.
+
+A server that has successfully transitioned to using 1-RTT keys could attempt to
+decrypt these packets, which will be unsuccessful.  Though a server that has
+1-RTT keys can safely discard any unprotected handshake messages from the
+client, an unprotected ACK frame could indicate that the server needs to
+retransmit its own handshake messages.  ACK frames therefore cannot be
+discarded.
+
+Until the server has received a positive acknowledgment for all of its
+unprotected handshake messages, either in the form of an explicit acknowledgment
+or implicitly through the use of 1-RTT keys, it MUST NOT discard packets with a
+KEY_PHASE of 0 if they cannot be decrypted successfully.  If these packets are
+unprotected and contain ACK frames, those ACK frames MUST be recovered and used
+to determine whether to retransmit TLS handshake messages.
+
+TBD:
+
+: We could observe/require that all unprotected packets are marked with a
```

Comment:
I like this logic.  I will make that change.

I suspect that servers will likely have very different tolerances for 0-RTT packets, particularly in the absence of the ClientHello. The ClientHello will probably be the only place that we can put a source address validation token.